### ISSUE TITLE
Rework lambda mechanism to improve speed and usability

### ISSUE THREAD

**Issue by porkybrain** [edouarda/brigand#177]
moved here to stop spamming twitter

Edit:

Relevant tweet: https://twitter.com/odinthenerd/status/740900401340354560

**Comment by porkybrain** [comment]
definition of terms:
selector = what `args` is now, the _1, _2 etc. tags
lambda = a template with place holder `selector`s in it or any of its children
aggregate = a template with no placeholders in its parameters or their parameters recursively
 
a lambda is basically a tree of types with either `selector` types or non template types at the leafs. we really don't have any way to tell if a template type we encounter is a `lambda` or just an `aggregate` so we test now. If we were to write a recursive function which walks the tree unpacking types and swapping out the selector types in such a way that it does not matter if it is walking a `lambda` or an `aggregate` we would not have to test and this could (my gut tells me) run faster.

we need a way of treating lambdas as types 'in current context' which we call deferr or protect or something else depending on the MPL now. If we were to specialize our walker to unwrap but terminate drill down on such types we could use that tag to minimize walking into branches which have no `selector`s (say my lambda is looking for a match to a old school type list, with no tag we have no way of knowing whether to drill down to the end of that list looking for a placeholder).

For that matter we could force the user to tag any non lambda argument, then drill down would be lightning fast and we wouldn't need `protect` which can be confusing to stupid users like me at first. maybe we could call it _p as in:
```C++
using l = list<list<int>, int, list<list<int>, int>>;  //don't want to walk this
using res = find < haystack, or_<
    std::is_same<_p<l>, _1>, 
    std::is_same<_p<std::is_pod<_1>>, _1> //we are looking for a match for is_pod, not executing it
>>;
```

**Comment by porkybrain** [comment]
```C++
namespace brigand{
	template<typename T>
	struct _p {};
	template<std::size_t N>
	struct selector{};

	namespace detail {
		template<typename T, typename...Ts>											//user error
		struct call_impl {
			using type = T;
		};

		template<template<typename...> class F, typename... Fs, typename...Ts>		//lambda
		struct call_impl<F<Fs...>, Ts...> {
			using type = typename F<typename call_impl<Fs, Ts...>::type...>::type;
		};

		template<std::size_t N, typename...Ts>										//placeholder (should be fast tracked)
		struct call_impl<selector<N>, Ts...> {										
			using type = at_c<list<Ts...>, N>;
		};

		template<typename T, typename...Ts>											//non lambda parameter
		struct call_impl<_p<T>, Ts...> {
			using type = T;
		};
	}

	template<typename T, typename...Ts>
	using call = typename detail::call_impl<T, Ts...>::type;						//calls a lambda 
}
using namespace brigand;
using t1 = call<_p<int>>;
using t2 = call<
	or_<
	selector<0>,
	_p<int16_t<0>>
	>,
	int16_t<0>>;
using t3 = call<
	or_<
	selector<0>,
	_p<int16_t<0>>
	>,
	int16_t<0>>;
using t4 = call<
	std::is_same<
		or_<
			selector<0>,
			int16_t<0>
		>,
		int16_t<0>
	>,
	int16_t<0>>;
```

**Comment by brunocodutra** [comment]
If I may weigh in, I totally agree with @porkybrain's view. The distinction between a *lambda expression* and a *list* based on the presence of a placeholder is very artificial. What if one needs a *constant lambda expression*, i.e., one which is bound to fixed parameters?

That said, I feel that providing support for *lambda expressions*, which are lazy by definition, doesn't make much sense given that the rest of Brigand is eager, but that's just my opinion of course.

**Comment by brunocodutra** [comment]
> Anything that gets us away from the nested apply template should be a step in the right direction, template members of template classes seem to be a dusty corner. 

Perhaps you should consider something along the same lines of Metal. It never relies on nested types/templates, rather an *invocable* is any type that matches the following signature (or will be after I merge brunocodutra/metal#32 which moves entirely to an eager design):

~~~{.cpp}
template<template<typename...> class> struct invocable{};
template<typename...> using f = /* ... */;

using result = metal::invoke<invocable<f>, args...>; // -> f<args...>
~~~

Internally `metal::invoke` extracts `f` out of `invocable<f>` by matching its signature.

**Comment by porkybrain** [comment]
So your are on the other side of the wrap lambdas vs. wrap non lambdas debate ;) somehow I just like the syntax `transform<l,std::add_pointer<_1>>` too much.

**Comment by brunocodutra** [comment]
> So your are on the other side of the wrap lambdas vs. wrap non lambdas debate

I'm not sure which side is which, but judging from your implementation above, I’d say I’m on your side. Basically I follow the Lisp notation, i.e., a *list* is an expression unless *quoted* (*protect* is called *quote* in Metal BTW).

> somehow I just like the syntax `transform<l,std::add_pointer<_1>>` too much

It is neat indeed and the sole reason why Metal hasn't gone eager long ago, but I finally gave up on *lambda expressions* in favour of the slightly more verbose, but strictly equivalent `transform<l, bind<lambda<std::add_pointer_t>, _1>>`

**Comment by jfalcou** [comment]
Multiple points:

 - Brigand is either lazy or eager. YOu have the choice brigand::foo is eager, brigand::lzay::foo is lazy. YOu need lazyness for higher order compositionnability

- I'm open to w/e break we need as long as the syntax is not horrible

**Comment by brunocodutra** [comment]
> YOu need lazyness for higher order compositionnability

I wouldn't go as far as to say lazyness is *required* for composability, but I agree it makes it terser.

**Comment by ericniebler** [comment]
Meta has lambdas but avoids them internally and discourages their use because they are too heavyweight. Meta prefers `transform<l, quote<std::add_pointer_t>>` to the lambda equivalent of `transform<l, lambda<_a, defer<std::add_pointer_t, _a>>>`

**Comment by porkybrain** [comment]
Totally agree with Eric, plus I like the bind stuff in meta. Actually though when adding pointers to 1000 unique internal_constants my call function is only 5% slower than metas  transform<l, quote<std::add_pointer_t>> so much faster that current brigand. 

The more I think about it though the more I think assuming eager will end up faster all be it uglier than assuming lazy. We may have to make a fundamental decision at some point how much noise we are willing to put up with in the name of speed. 

I will try to get some more bench marks done, hard to concentrate at home though, my son is teething.

**Comment by edouarda** [comment]
My work here is done, I can now resume hitting F5 on the stars counter.

**Comment by porkybrain** [comment]
I updated my algorithm (see second post in this thread) adding a few fast tracks and now its beating meta (by a hair) on clang 3.7 using real lambda syntax (at least in my test). 
For my (very unscientific) test I have a 1000 element list of unique intigral_constants which gets transformed in different ways. I comment in only one of the tests below and use the compiler diagnostic to tell the compilation time (I cranked up verbosity). 

```C++
using a = meta::transform <l, meta::bind_front<meta::quote<is_same_t>, float>>;
using b = transform<l, std::is_same<_1,_p<float>>>;
using c = meta::transform < l, meta::quote<std::add_pointer_t> >; 
using d = transform<l, std::add_pointer<_1>>; 
``` 

Let me know if there is a faster way to do *a*, I'm not good at meta yet. In both cases meta was a few percent slower. Maybe we don't need binding after all? (or maybe a overlooked something, happens often enough ;))

**Comment by edouarda** [comment]
Maybe we can improve the lambda plumbing to make it faster?

**Comment by porkybrain** [comment]
I think the _p<> is key, it keeps us from having to walk through types which are not lambas just to make sure they aren't and more importantly allows us to fast track in the common trivial case of one parameter. That and getting rid of the nested apply template. For example *b* is like 10x faster with my new algorithm than with current brigand.

**Comment by edouarda** [comment]
Why "_p" ? What does it mean?

**Comment by edouarda** [comment]
You are not protecting the parameter, you are hinting brigand that it's not a lambda. In other way, you are pinning the parameter, what about brigand::pin<> ?

**Comment by porkybrain** [comment]
the idea would be if you use _p on a lambda it would have the same effect as brigand::protect has now, so there is no need for many different flavors of protection.

**Comment by porkybrain** [comment]
Its probably not the right place to argue on behalf of sexiness but I would like to have the power of meta lambdas but the terseness of `transform<l, std::add_pointer<_1>>`, have it all be the same thing, at least conceptually and have my cake and eat it too, and a puppy for that matter. 

Jokes aside maybe we could make an argument alias for the case where you need to reference a lambdas arg from a nested lambda, when walking we could just tack on the extra args to the args we already have and have the alias be a special selector. Problem with that is we would have to start wrapping things again which kills the sexiness. This would also screw up short circuiting but only in the case that it is used. 

We could make a 'super<_1>' tag which would select the arg from its "superlambda" as in one level up, and pass all the args upstream while walking, this would make the carteasan product look like this

```C
template <class L>
using cart_prod = reverse_fold<
    L, list<list<>>,
    l::join<l::transform<
            _2, _p<l::join<l::transform<super<_1>, p<list<l::push_front<_1, super<_1>>>>>>>>>>>;

```

**Comment by brunocodutra** [comment]
Regarding the implementation of the `cart_prod`, wouldn't it be better to nest calls to `transform`?

That's how it would look like using lambda expressions in current [lazy] Metal

~~~{.cpp}
template<typename x>
using cart_prod = apply_t<
  lambda<join>,
  transform_t<
    transform<list<quote<_1>, quote_t<_1>>, quote_t<x>>, 
    x
  >
>;
~~~

Which I admit is much more readable than the equivalent implementation in [eager] Metal as of currently standing brunocodutra/metal#32

~~~{.cpp}
template<typename x>
using cart_prod = apply<
  lambda<join>,
  transform<
    bind<lambda<transform>, bind<lambda<bind>, quote<lambda<list>>, bind<lambda<quote>, _1>, quote<_1>>, quote<x>>, 
    x
  >
>;
~~~

Well, that's the price I'll have to pay for refusing to support an eager/lazy hybrid.

**Comment by porkybrain** [comment]
yes there do seem to be trade offs between different implementations of MPLs which is why I called attention to the fact that some hybrid of current MPLs will no doubt be standardized and we really should get that right. Lambdas do seem to be the crux too, as far as I can see standardizing a public interface of most algorithms does not dictate their implementation and is therefore not so important, with lambdas it is... 
I do think speed is pretty important because everyone who does not need speed will be using hana style. Library implementer and infrastructure builders need speed at all cost. I would like to establish a baseline of how fast we can make sexy run before we start making it ugly in the name of speed.
By the way although I seldom agree with you I am very glad there are others with different views out there. What is the advantage to being all eager?

**Comment by edouarda** [comment]
Speed is paramount and one of the main goal of brigand. Because meta programming libraries are a fundamental building blocks, there is an amplification effect of their compilation speed. If we deliver a lightweight, instant compile time metaprogramming library, library will be more open minded to include metaprogramming into their arsenal.

**Comment by edouarda** [comment]
I agree however we should make it easy of use first, then make it fast, to the limit it has to remain comfortable to use.

**Comment by ericniebler** [comment]
I suspect you'll find that if you try to standardize meta-lambdas at all, you'll make your job 10x harder.

**Comment by edouarda** [comment]
I agree with @ericniebler I think we should bring a solution that's convenient for most use cases. If I grep quasardb source code, we have little to no complex meta-lambda.

**Comment by porkybrain** [comment]
basically every algorithm takes a lambda/callable, if you don't standardize the 'callable' interface then whats left to standardize in an MPL?

**Comment by ericniebler** [comment]
In meta, `Callable` is just a class type with a nested `invoke` template alias. It doesn't need to be more complicated.

**Comment by edouarda** [comment]
@porkybrain I think we shouldn't obsess with standardization but more about making the library useful. Standardization will appear as the libraries mature.

**Comment by porkybrain** [comment]
It makes it so that you have to wrap everything that does not have an invoke template which pollutes the API and I'm still not convinced its the fastest solution out there.

**Comment by ericniebler** [comment]
Template template parameters are the fastest solution or there. Aside from that, sprinkling meta::quote here and there seems no worse to me than sprinkling brigand::_p here and there.

**Comment by brunocodutra** [comment]
> By the way although I seldom agree with you I am very glad there are others with different views out there. What is the advantage to being all eager?

Being either all eager or all lazy, doesn't matter which, has the great advantage of keeping the API concise and symmetrical. See, users have a very limited attention span (speaking from personal experience) and it is very important that a library be easy to grasp (especially if you think about standardization). Metal for instance is entirely built upon 5 very intuitive and straightforward concepts, which can be defined by no more than a couple of lines of text each: `Value`, `Number`, `Lambda`, `List`, `Pair` and `Map`.

Now regarding the verbosity of eager composition that I mentioned before, I found a very simple solution that does not require lazy metafuncions and I can't believe it didn't occur to me before. One only needs to provide alias such as the following (notice the Capital letter)

~~~{.cpp}
template<typename lbd, typename... seqs>
using Transform = bind<lambda<transform>, lbd, seqs...>;

template<typename lbd, typename... lbds>
using Bind = bind<lambda<bind>, lbd, lbds...>;

/* ... */
~~~

And voilà! We have the terseness of lambda expressions without a single lazy metafunction

~~~{.cpp}
template<typename x>
using cart_prod = apply<
  lambda<join>,
  transform<
    Transform<Bind<quote<lambda<list>>, Quote<_1>, quote<_1>>, quote<x>>, 
    x
  >
>;
~~~


> EDIT: in fact, in Peter Dimov's world, it's even simpler. Just pass template template parameters.

But then you lose the ability to compose and along with it the joy of TMP.

**Comment by brunocodutra** [comment]
> Template template parameters are the fastest solution or there. 

@ericniebler That's the approach of [eager] Metal, but in order to retain the ability to compose, they are simple wrapped within a type of appropriate signature:

~~~{.cpp}
template<template<typename...> class> struct wrapper {};

using callable = wrapper<std::add_pointer_t>; // anything that matches this signature is callable
~~~

**Comment by ericniebler** [comment]
> I found a very simple solution that does not require lazy metafuncions and I can't believe it didn't occur to me before. One only needs to provide alias such as the following ...

You've discovered `meta::defer`. :-) Everything in Meta's `lazy` namespace is a type alias defined in this way with `defer`.

> in order to retain the ability to compose, they are simple wrapped within a type of appropriate signature

In Meta, `wrapper` is called `quote`. We keep rediscovering the same things over and over.

**Comment by brunocodutra** [comment]
> You've discovered meta::defer. :-) Everything in Meta's lazy namespace is a type alias defined in this way with defer.

Hmm, please correct me if I'm wrong, but when I last took a look at Meta, it seemed to me that `meta::defer` was rather a lazy adaptor for eager metafunctions, i.e., conditionally defining `::type` to be the result of the invocation so as to make sure it is SFINAE friendly. What I propose is somewhat different, as there won't be any nested `::type` anywhere. In my example above `Transform<F, L>::type` doesn't mean anything as it is **noy** a lazy metafunction, but only an adaptor from lambda expressions to strictly equivalent, but much terser, bind expressions.

> In Meta, wrapper is called quote. We keep rediscovering the same things over and over.

Indeed, I'd noticed it before, the difference is that Meta relies on `meta::quote` to define a nested `invoke<>` much like MPL, whereas Metal allows *any* type that matches that very signature to be *invoked* by simply  *extracting* the metafunction by template matching at the point of instantiation. BTW, I compared extracting the metafunction from the type signature against relying on a nested template and I seem to have observed some performance penalty for the latter both on GCC and on Clang. In fact I observe the same thing when accessing any nested element of a type, *especially* `std::integral_constant`'s `::value`, when compared to matching its template signature to extract the value directly.

**Comment by pfultz2** [comment]
> Hmm, please correct me if I'm wrong, but when I last took a look at Meta, it seemed to me that meta::defer was rather a lazy adaptor for eager metafunctions, i.e., conditionally defining ::type to be the result of the invocation so as to make sure it is SFINAE friendly.

You could create a `defer` alias that would do this, then:

```cpp
template<template<class...> class Template, class... Xs>
using defer = bind<lambda<Template>, Xs...>;
```

It is terser than having to write `bind<lambda<transform>, lbd, seqs...>`.

**Comment by brunocodutra** [comment]
> You could create a defer alias that would do this, then:

Indeed, I thought about that as well, but I found it didn't help very much on the readability side.

~~~{.cpp}
template<typename x>
using cart_prod = apply<
  lambda<join>,
  transform<
    defer<transform, defer<bind, quote<lambda<list>>, defer<quote, _1>, quote<_1>>, quote<x>>, 
    x
  >
>;
~~~

Perhaps I should provide it for easy integration of user defined metafunctions, but still provide *lambda expression friendly* aliases  on top of `defer` for every construct in Metal. It's literally just a couple of lines per metafunction anyways.

~~~{.cpp}
template<typename lbd, typename... seqs>
using Transform = defer<transform, lbd, seqs...>;
~~~

Back to the topic, call me obsessive compulsive, but I'm now totally convinced a hybrid implementation is both wonky **and** entirely unnecessary. Just my 2 cents.

**Comment by ericniebler** [comment]
> Perhaps I should provide it for easy integration of user defined metafunctions, but still provide lambda expression friendly aliases on top of defer for every construct in Metal.

This is what meta does.

> I compared extracting the metafunction from the type signature against relying on a nested template and I seem to have observed some performance penalty for the latter both on GCC and on Clang.

Mere mentions of an instantiation (via an alias) are less expensive than mentions of a nested identifier, since that forces the instantiation of the enclosing type. That is, X<T> is cheaper than X<T>::Y.

**Comment by brunocodutra** [comment]
>>    Perhaps I should provide it for easy integration of user defined metafunctions, but still provide lambda expression friendly aliases on top of defer for every construct in Metal.
>
> This is what meta does.

Cool! I guess we somewhat agree on the API in general, albeit not so much on the implementation. 

It seems to me that if we were to regard Meta's use of nested `invoke<>` and `type` strictly as hidden implementation details, then Meta and [eager] Metal should look very similar from the user's perspective. On the other hand I posit that if you try to document these as part of Meta's customization points, you'll find it rather challenging to summarize the API in a concise and symmetrical manner.

@porkybrain Why instead of implementing lambda expressions directly, don't you simply provide *lambda expression friendly* aliases just like Metal?

~~~{.cpp}
namespace lazy {
    template<typename... seqs>
    using append = bind<quote<brigand::append>, seqs...>;

    template<typename L, typename F>
    using tranform = bind<quote<brigand::tranform >, L, F>;

    /*...*/
}
~~~

By doing this you can write lambda expressions out of the box using the existing mechanism for callable invocations.

**Comment by brunocodutra** [comment]
>>  I compared extracting the metafunction from the type signature against relying on a nested template and I seem to have observed some performance penalty for the latter both on GCC and on Clang.
>
> Mere mentions of an instantiation (via an alias) are less expensive than mentions of a nested identifier, since that forces the instantiation of the enclosing type. That is, X is cheaper than X::Y.

What I meant to say is that

~~~{.cpp}
template<typename N>
using not_ = bool_<!N::value>;
~~~

seems to be much slower than 

~~~{.cpp}
template<typename N>
struct not_impl {};

template<typename T, T v>
struct not_impl<std::integral_constant<T, v>> : std::integral_constant<bool, !v> {};

template<typename N>
using not_ = typename not_impl<N>::type;
~~~

despite the fact both internally mention a nested entity.

**Comment by ericniebler** [comment]
That's surprising if it's true. I wonder what's going on there.

I'll note that the latter implementation is broken by code like the following:

```
struct MyFalse {
  using type = MyFalse;
  static constexpr bool value = false;
};

using MyTrue = not_<MyFalse>; // ERROR
```

That is too fragile for my taste.

**Comment by brunocodutra** [comment]
> That's surprising if it's true. I wonder what's going on there.

I think the difference is that by accessing the nested `::value` of an `std::integral_constant` you force the instantiation of a type which defines lots of non-trivial nested entities, whereas by matching the type signature you only force the instantiation of `not_impl` which defines a single nested `::type` and nothing else.

>  I'll note that the latter implementation is broken by code like the following:

I should have mentioned that a `Number` in Metal must be an instance of `std::integral_constant`, i.e. `MyFalse` in your example above is **not** a `Number` in Metal parlance. As a matter of fact, every concept in Metal requires a very specific type signature, which allows concept checking by partial template specialization and makes it rather trivial to provide traits such as `is_number`, `is_map`, etc.

**Comment by ericniebler** [comment]
> As a matter of fact, every concept in Metal requires a very specific type signature, which allows concept checking by partial template specialization and makes it rather trivial to provide traits such as `is_number`, `is_map`, etc.

Forgive the rather impolite question, but ... if you make these design compromises to make Metal fast, then why is it so slow? http://ldionne.com/metabench/

**Comment by brunocodutra** [comment]
>> As a matter of fact, every concept in Metal requires a very specific type signature, which allows concept checking by partial template specialization and makes it rather trivial to provide traits such as is_number, is_map, etc.
> 
> Forgive the rather impolite question, but ... if you make these design compromises to make Metal fast, then why is it so slow? http://ldionne.com/metabench/

Don't wrorry, that's not an impolite question, since I don't strive to beat the competition in terms of performance. I don't make those design choices to make Metal fast, it's the other way around, I make them in order to allow strict concept checking. It just so happens that in this particular case this seems to be the fastest alternative, at least in my experience.

Now to answer your question I'll have to dive a bit into the internals of Metal. Basically, every algorithm in Metal that can’t be expressed in terms of `transform` is expressed in terms of `fold`. Now the usual implementation of `fold` implies a linear growth of the template instantiation depth, which is limited to 256 on Clang and to 900 on GCC by default and thus requires the user to provide special flags in order to go beyond that limit. I didn't want that. Another issue with that approach is that memory also grows very rapidly. I know that could be worked around by using fast-tracking, but that’s something I refuse to do in the name of code clarity and maintainability.

Instead, I [chose to implement `fold`](https://github.com/brunocodutra/metal/blob/bare/include/metal/list/fold.hpp) in a way that implies **logarithmic** growth of the template instantiation depth. Basically it keeps track of two indices `beg` and `end` and recursively folds each half of the list. Another advantage of this implementation is that you also have `reverse_fold` by chosing `beg` to be greater than `end` and `partial_fold` by chosing the indices to be a sub-range of the list. Now the problem with this approach is that it necessarily relies on `at` to actually access each and every element, since it only keeps track of indices. That is what hurts the performance so bad, especially because core algorithms such as `join` must rely on `fold`. 

At least it turns out this approach is much less memory hungry. Perhaps Metal will look better on Metabench once we manage to provide memory benchmarks.

**Comment by porkybrain** [comment]
@brunocodutra if everything relies on fold I don't see the harm in fast tracking it, surly fast tracking one algorithm will not hurt maintainability so much that it outweighs the speed advantage.

**Comment by edouarda** [comment]
Our users want speed and fast tracking delivers that.

**Comment by brunocodutra** [comment]
There's no arguing fast-tracking delivers speed and of course [it seems reasonable to back core algorithms by it](https://github.com/brunocodutra/metal/issues/14), however so far I've been reluctant to do so for many reasons:

* I really like the logarithmic recursion. It is elegant, generic and unique among present day TMP libraries.
* It eats less memory.
* So far I've been focusing on designing the perfect API and I spare no complete rewriting if I think I can make things the slightest more elegant, so don’t want implementation details to get in my way.
* I don’t want to pay for what I don’t need and so far Metal is fast enough for its use cases.
* Brigand and Meta already do a good job on delivering speed, if Metal goes fast-tracking it will lose its identity and become just another face in the crowd.
* My design choice regarding `fold` is the most emblematic regarding the *elegance* X *speed* dilemma, but of course there are other. Notably, I always prefer to reuse algorithms to implement new ones so that out of the dozens Metal provides the vast majority are two liners that leverage on the few (< 10) core algorithms. If `fold` goes fast-tracking, the next logical question will be why implementing `push_back` and `push_front` in terms of `join` if they can be implemented directly just as easy (albeit more verbosely) and overnight Metal triples in size. It’s a Pandora’s box and **that’s** what I mean when I refer to maintainability.

Please don't get me wrong, I understand Brigand has its use cases and sheer speed might be required, I just don't agree this is the only logical approach to the TMP problem.

**Comment by porkybrain** [comment]
I have been testing my lambda idea against meta with clang and 100 unique instantiations of this tree:
`using t0 = meta::list<meta::list<i<10>, i<20>>, meta::list<i<30>, i<40>>>;`

first test just adding a pointet to every sub list, does not use meta::lambda, lightning fast:
meta 50ms
```
template<typename T>
using tr = transform < T, quote<std::add_pointer_t> >; 
```

new brigand 50ms
```
template<typename T>
using tr = transform<T, std::add_pointer<_1>>;
```

I actually can't claim I understand the reason why these are equivalent, seeing as new brigand has more instantiations, on the other hand it does not use any nested template anything (meta uses a nested invoke alias). My gut would tell me meta should be faster here but even if I use other functions besides `add_pointer` and increase the size or amount of trees by 10x the two seem identical or at least within the noise of each other.

now how about adding pointers to the i's nested in the sub lists (cascading):

meta using lambda 590ms (can I write this in a faster way?)
```
namespace l = meta::lazy;
template<typename T>
using tr = transform<T, lambda<_a, l::transform< _a, quote<std::add_pointer_t>>>>;
```
my new brigand lambda equivalent 150ms (nested lambdas that can use parameters of their parents are wrapped with `_l`, in this case we are not actually using the parent args)
```
template<typename T>
using tr = transform<T, lazy::transform<_1, _l<std::add_pointer<_1>>>>;
```

to test cascading lambdas I wrote my own do nothing metafunction:
```
template<typename T, typename U>
struct myLazyF {
	using type = i<(meta::size<T>::value + U::value)>;
};
template<typename T, typename U>
using myF = typename myLazyF<T, U>::type;
```

meta lambda actually using parents args 1710ms
```
template<typename T>
using tr = transform<T, lambda<_a, l::transform< _a, lambda<_b, defer<myF, _a, _b>>>>>;
```` 

new brigand lambda equivalent 300ms (parent args must be wrapped with `parent<>`, `parent<parent<parent<...>>>` is possible for the freaks among us)
```
template<typename T>
using tr = transform<T, lazy::transform<_1, brigand::_l<myLazyF<parent<_1>, _1>>>>;
```

In conclusion I think a unified syntax might be possible where brigand is the same speed in cases where meta uses "invokables" for speed but also has the flexibility of meta lambdas (with a considerable performance edge).

**Comment by edouarda** [comment]
That sounds pretty awesome!

**Comment by jfalcou** [comment]
Do we have a migration plan then ?

**Comment by porkybrain** [comment]
I found a case for meta, if I write my own struct which contains an invoke and that invoke does not need to call a metafunction as in:
```
struct add_pointer_f {
	template<typename T>
	using invoke = T*;
}; 
```
meta is faster. However I would argue that that is probably a corner case as it is only possible when wrapping a type or adding type qualifiers to it. In all other cases you need specialization of some kind or another when means an alias alone will not work. 

pattern matching `F<_1>` can actually be done at algorithm level, matching for something that has an "invoke" is not possible. Therefore we could optimize for the "common case" of F<_1> to try and make the gap smaller. 

I still want to get this right before we do any migration, I think there may be a way to wrap meta invokables in a way that we can still profit from their speed. Plus my implementation probably has some corner cases where it has bugs right now. 

I would also like to get variadic meta lambdas working (as in _args rather than _1), however I'm not sure if I really know how to get the public interface to not suck.

**Comment by porkybrain** [comment]
Found another tweak to 'new brigand' model, now seems to be neck and neck again.

What is faster seems to be super benchmark dependent. Benchmarking with 1000 transforms processing (adding pointer to) only one unique element each they are super similar. However when processing 1000 lists of 4 unique elements each the results are interesting:
```
template<template<typename...> class F>
struct magic{};

template<typename T>
struct classicFunctor {
	using type = T*;
};

template<typename T>
using magicFunctor = T*;

struct metaFunctor {
	template<typename T>
	using invoke = T*;
}; 


meta::transform < T, metaFunctor >;     //  220 
meta::transform < T, meta::quote<classicFunctor> >;     //  660  
brigand::transform<T, classicFunctor<_1>>;    //180 
brigand::transform<T, magic<magicFunctor>>;      // 140 
```
"magic" is an alias wrapper (inspired by metal) which we will probably name quote (or turbocharge?) or something and can coexist with the "new brigand" strategy. 
I think that this shows that at least on my machine with clang 3.7 that meta is beatable with a unified lambda syntax.

**Comment by porkybrain** [comment]
We could name "magic" "eager" and also allow "eager" to wrap the lambda in the cascading lambda use case if args match as in:
```
template<typename T>
using tr = transform<T, lazy::transform<_1, eager<std::add_pointer_t>>>;
```

**Comment by ericniebler** [comment]
I know meta::quote is implemented in terms of defer because of a core issue. There is a comment in the code. defer makes quote instantiate more templates than it would otherwise need to. The core issue only surfaces in certain situations. (I could quickly find the problem again by removing the workaround and seeing what breaks.) I'm curious whether brigand suffers from the same problem and will need a similar workaround. I would check but I'm not at my computer.

**Comment by ericniebler** [comment]
"core issue" == "core language issue"

**Comment by porkybrain** [comment]
yes I was expecting that was the reason for 'meta::transform < T, meta::quote<classicFunctor> >' being so slow. 
I also expect that we will find bugs or corner cases in my implementation so no cracking champagne yet ;) 
I am coming around to the idea of an all eager interface as suggested by @brunocodutra although It does fuck with your mind at first glance. If we were to rename 'eager' again to 'call' and give it a signature like meta::defer (as in template<template<typename...> class F, typename...Args>) it would result in something like this which doesn't look half bad:
```
template<typename T>
using tr = transform<T, call<transform, _1, call<std::add_pointer_t, _1>>>;
```

**Comment by brunocodutra** [comment]
The *core issue* shows up specificaly when instantiating an alias template
with variadic arguments in place of non-variadic parameters outside of a
SFINAE context. Metal works around it by avoiding instantiating things
within **nested** alias templates, hence the pattern matching to extract
the metafunction. Otherwise a double instantiation within a SFINAE context
is required.

"core issue" == "core language issue"

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/edouarda/brigand/issues/177#issuecomment-227164909>,
or mute
the thread
<https://github.com/notifications/unsubscribe/AD2zj5omYAFC8G4gMhoodvMHcnHqYRLjks5qNqiEgaJpZM4Iy6W9>
.

**Comment by porkybrain** [comment]
Can't seem to get my head around transforming something like this:
```
template<typename T>
using tr = transform<is_same<remove_reference<T>, conditional<is_same<int,add_pointer<T>,bool,add_pointer<_1>>>>; 

```
into only eager... maybe keep hybrid?

[core language issue]
I think 'new brigand' is not affected because I only do pattern matching specialization, no nested stuff (its slow any way).

**Comment by brunocodutra** [comment]
In Metal I'd make it lazy by replacing /metafunction</ by
/bind<lambda<metafunction>, /, which can be made terser by the alias
templates I suggested before.

Brigand's bind doesnt appear to allow this as easily, because the
MetafunctionClass being bound isn't implicit protected, so I guess it would
require some tweaking.

**Comment by brunocodutra** [comment]
I mean, I would make the eager equivalent lazy by following the rule above.

**Comment by porkybrain** [comment]
if we had wrappers 'call' and 'lcall' (lazy call) which also implicitly protect the user could decide on a lambda by lambda basis which syntax form they want and we could get rid of the 'protect', 'magic', '_l' and whatnot I have been using. That way we could still pattern match aliases because they would be wrapped in a 'call' at top level and have the speed that comes with it (equivalent to 'magic' in the last benchmark). We could assume 'lcall' at top level to make things terser (assuming 'call' is not possible) Following this idea @ericniebler Cartesian product would look like this:
```
template <class L>
using cart_prod = reverse_fold<
    L, list<list<>>,
    l::join<l::transform<
            _2, lcall<l::join<l::transform<parent<_1>, lcall<list<l::push_front<_1, parent<_1>>>>>>>>>>>;
```
add_pointer would look like this:
```
template<typename T>
using tr = transform<T, call<std::add_pointer_t>>; //probably fastest possible implementation
//or
template<typename T>
using tr = transform<T, std::add_pointer<_1>>; //slightly slower
```
we would still need to tag things that are not lambdas but we could call that 'pin' rather than '_p':
```
remove_if<T,is_same<pin<int>,_1>>
```
parameters that are not lambdas actually don't happen as often as one might think looking at my kvasir.io code base.

**Comment by ericniebler** [comment]
> The *core issue* shows up specificaly when instantiating an alias template with variadic arguments in place of non-variadic parameters outside of a SFINAE context. Metal works around it by avoiding instantiating things within **nested** alias templates, hence the pattern matching to extract
the metafunction. Otherwise a double instantiation within a SFINAE context is required.

Interesting! Meta's choice of a nested `invoke` alias would seem to necessitate this expensive `defer` workaround that wouldn't otherwise be necessary. That's something to chew on.

For the record, I'm aware that meta's lambda mechanism is heavyweight. It's there as a curiosity, not because I think it's useful or necessary. I didn't believe an MPL-like lambda syntax could be provided without an MPL-like perf hit. If that's not true, it changes things.

I'm not wild about the need for `pin`. What about the design necessitates it? Can't the lambda evaluation mechanism simply interpret any type it can't deconstruct via pattern matching as a literal type and substitute it directly? Then `pin` is only needed as an optimization to keep the lambda mechanism from recursing into a particular template instantiation.

**Comment by porkybrain** [comment]
yeah pin seems to be the thing people don't like. 

Its not just about stopping recursion (as I originally falsely claimed), if you evaluate everything that is not a pin, a call or an lcall as a lazy function then everything is just pattern matching, no SFINAE tests or anything needed. I am not aware of a non SFINAE test which shows whether a type has a nested type alias (i.e. is a lazy function). Besides the user may want to pass a function as a type so we need something like pin any way. You could alternatively wrap every function but that is bloating the common case rather than the uncommon case. I have not tested the speed of using a SFINAE test but it seems to also be a teachability problem if the use of pin relies on whether a type you want to pass as a type happens to have a nested type alias.

**Comment by porkybrain** [comment]
It is trivial to make everything that does not mach ` template<template<typename...> class F` be interpreted as a type, that seems hard to teach though.

**Comment by brunocodutra** [comment]
That's what Metal does and TBH I don't see how this would be harder to
teach than by evoking the presence of some nested template.

**Comment by porkybrain** [comment]
```
transform<T, std::is_same<_1,std::array<int,4>> //works
transform<T, std::is_same<_1,std::false_type>> //works
transform<T, std::is_same<_1,std::vector<int>> //error
transform<T, std::is_same<_1,std::string>> //error
```
this does not seem that intuitive.

**Comment by porkybrain** [comment]
coming back to @pdimov 's world view, adding a wrapper around template template arguments (which I called `call` above) is less of a perf hit as long as you extract it using pattern matching. It is still faster on clang (your mileage may vary) than a nested alias.

meta with custom functor with pure 220ms
new brigand 180 ms
new brigand call wrapper pure alias 140ms
dimov impure alias 140ms
dimov pure alias 90ms

remember "pure aliases" (where you never need to retrieve a nested type) are the uncommon case so its really 140 vs. 180. That penalty buys you things like overloading (transform would have to take the function first in @pdimov 's world our not be capable of two lists), composition and binding (without creating aliases yourself) and 'generic' algorithms (no non types in public interface). 

Having algorithms take non type parameters just seems wrong to me, but who knows maybe I'm just being dogmatic. What would Cartesian product look like in @pdimov style?

**Comment by porkybrain** [comment]
Looking through ranges v3 [here](https://github.com/ericniebler/range-v3/blob/a4c64b55301b839dde7cbf7d40904c1ff5887320/include/range/v3/utility/concepts.hpp#L114) you would need to use pin 
```
meta::apply<
                    meta::quote<meta::lazy::strict_and>,
                    meta::transform<
                        base_concepts_of_t<Concept>,
meta::bind_back<meta::quote<concepts::models>, Ts...>>>;
```
but I still think 
```
wrap<
    transform<
        base_concepts_of_t<Concept>,
        call<concepts::models,_1,pin<Ts>...>>,
    strict_and>;
```
is sexier

**Comment by brunocodutra** [comment]
@porkybrain I think there was a typo in a previous message of yours, which led to some miscommunication. You mentioned any type that does not match `template<template<typename...> class F` is considered a regular type (i.e. not a lambda). That is clearly a typo, and I supposed you meant `template<template<typename...> class> class F` (notice the trailing `> class`), but now I realize you meant `template<typename...> class F` instead (notice the missing leading `template<`).

To clarify matters, Metal only consider types that match `template<template<typename...> class> class F` to be considered callables. That's what I claim to be intuitive and easy to teach.

**Comment by brunocodutra** [comment]
I see that I might not be very clear when I claim lazy semantics may be expressed using an all-eager implementation, so let me clarify the design:

Let `f` and `gs...` be any eager metafunctions, while `F = lambda<f>` and `Gs = lambda<gs>...` are types that match `template<template<typename...> class> class`.

Let `invoke<F, args...> := f<args...>` and `invoke<Gs, args...>... := gs<args...>...`.

Now assume `bind` is implemented so that `invoke<bind<F, Gs...>, args...> := invoke<F, invoke<Gs, args...>...>`, which is thus equivalent to `f<gs<args...>...>`.

Finally assume lazy versions of algorithms are defined as follows

~~~{.cpp}
namespace lazy {
    template<typename L>
    using reverse = bind<lambda<::reverse>, L>;

    template<typename X, typename Y>
    using equal = bind<lambda<::equal>, X, Y>;

   /* ... */
}
~~~

Using lazy aliases defined as such, `is_palindrome` could be expressed exactly as in MPL, without the need for actual lazy metafunctions.

~~~{.cpp}
using is_palindrome = lazy::equal<_1, lazy::reverse<_1>>;

static_assert(invoke<is_palindrome, list<int, void, void, int>>{}, "");
~~~

**Comment by porkybrain** [comment]
Sounds like I need to have a look at your design more closely. I came down with a cold today so my IQ dropped by like half since this morning so I'm not making any promises ;)

**Comment by edouarda** [comment]
@porkybrain perhaps you are overthinking this. It's probably better to iterate on designs than to try to come up with a "perfect" solution.

**Comment by porkybrain** [comment]
Hehe I don't want to be the guy that broke brigand twice :-) 

sometimes its good to get down to first principals. I'm not sure I'm there yet.

**Comment by pdimov** [comment]
> What would Cartesian product look like in @pdimov style?

https://github.com/pdimov/mp11/blob/master/include/boost/mp11/algorithm.hpp#L205
https://github.com/pdimov/mp11/blob/master/test/mp_product.cpp

**Comment by porkybrain** [comment]
Ok so basically mp11 product is hand coded for speed... I guess it all comes down to our expected user and use case. We will never be as fast with lambdas as the hand coded equivalent so it all comes back to the speed vs. Readability/writability/teachability trade off. 

A type wrapped template, which call essentially is, seems to be the next fastest thing possible. We can add an overload accepting call without breaking brigand. Seems to be a good first step in order to give trivial stuff, which is the common case, more speed.

**Comment by porkybrain** [comment]
this is getting hopelessly long, moving all 'call' related conversation to the pull request #185, and all lazy/cascading/variadic stuff to #186
